### PR TITLE
Adding Uniform distribution to PyTorch

### PR DESCRIPTION
**Pull request by neerajprad**
The implementation is straightforward. A couple of notes:
 - In our discussions, we had discussed throwing an error if `log_prob` got a value outside the distribution's support. This returns `-inf` instead. I thought that throwing an error does not seem to be a good response for a perfectly valid query. For reference, Tensorflow return `-inf` for uniform but throws an error for Bernoulli (scoring `[1]` with `Bernoulli([0.0])` for example). We should be consistent with whatever convention we choose.
 - Modified `sample_n()` to use `torch.Size` instead of a tuple `(n,)`. The reason is that `tensor.new(tuple)` just takes the last dim, as compared to `tensor.new(torch.Size)`. Not sure if that's expected. cc. @apaszke. 

```
In [7]: torch.zeros(2, 2).new((3, 2)).size()
Out[7]: torch.Size([2])

In [8]: torch.zeros(2, 2).new(torch.Size((3, 2))).size()
Out[8]: torch.Size([3, 2])
```

### CODE REVIEW

File: torch/distributions/uniform.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import torch
+from torch.autograd import Variable
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class Uniform(Distribution):
+    r"""
+    Creates a uniform random distribution parameterized by
+    `low` and `high`.
```
Comment (fritzo) [inline torch/distributions/uniform.py]: You might mention that return value is in the half-open interval `[low, high)`

File: torch/distributions/uniform.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        rand = self.low.new(shape).uniform_()
+        return self.low + rand * (self.high - self.low)
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        lb = value.ge(self.low).type_as(self.low)
+        ub = value.lt(self.high).type_as(self.low)
+        return torch.log(lb.mul(ub)) - torch.log(self.high - self.low)
```
Comment (fritzo) [inline torch/distributions/uniform.py]: nice!

**Comment by neerajprad** [comment]
Opened a PR in [pytorch#4328](https://github.com/pytorch/pytorch/pull/4328)
